yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -0,0 +1,515 @@
+    );
+    this.enqueueJob(stateIdentifier, PROVING_JOB_TYPE.ROOT_PARITY, () =>
+      this.runRootParityCircuit(rootParityInputs, stateIdentifier),
+    );
+  }
+
+  // Runs the root parity circuit ans stored the outputs
+  // Enqueues the root rollup proof if all inputs are available
+  private async runRootParityCircuit(inputs: RootParityInputs, stateIdentifier: string) {
+    const [duration, circuitOutputs] = await elapsed(() =>
+      executeRootParityCircuit(inputs, this.simulator, this.prover, logger),
+    );
+    logger.debug(`Simulated root parity circuit`, {
+      eventName: 'circuit-simulation',
+      circuitName: 'root-parity',
+      duration,
+      inputSize: inputs.toBuffer().length,
+      outputSize: circuitOutputs.toBuffer().length,
+    } satisfies CircuitSimulationStats);
+    if (!this.provingState?.verifyState(stateIdentifier)) {
+      return;
+    }
+    this.provingState!.finalRootParityInput = circuitOutputs;
+    this.checkAndExecuteRootRollup(stateIdentifier);
+  }
+
+  private checkAndExecuteRootRollup(stateIdentifier: string) {
+    if (!this.provingState!.isReadyForRootRollup()) {
+      logger('Not ready for root');
+      return;
+    }
+    this.enqueueJob(stateIdentifier, PROVING_JOB_TYPE.ROOT_ROLLUP, () =>
+      this.runRootRollup(
+        this.provingState!.getMergeInputs(0)!,
+        this.provingState!.finalRootParityInput!,
+        stateIdentifier,
+      ),
+    );
+  }
+
+  private storeAndExecuteNextMergeLevel(
+    currentLevel: bigint,
+    currentIndex: bigint,
+    mergeInputData: [BaseOrMergeRollupPublicInputs, Proof],
+    stateIdentifier: string,
+  ) {
+    const result = this.storeMergeInputs(currentLevel, currentIndex, mergeInputData);
+
+    // Are we ready to execute the next circuit?
+    if (!result.ready) {
+      return;
+    }
+
+    if (result.mergeLevel === 0n) {
+      this.checkAndExecuteRootRollup(stateIdentifier);
+    } else {
+      // onto the next merge level
+      this.enqueueJob(stateIdentifier, PROVING_JOB_TYPE.MERGE_ROLLUP, () =>
+        this.runMergeRollup(result.mergeLevel, result.indexWithinMergeLevel, result.mergeInputData, stateIdentifier),
+      );
+    }
+  }
+
+  /**
+   * Process the job queue
+   * Works by managing an input queue of proof requests and an active pool of proving 'jobs'
+   */
+  private async processJobQueue() {
+    // This is temporary to keep benchmarks consistent
+    const maxConcurrentJobs = 1;
+    // Used for determining the current state of a proving job
+    const promiseState = (p: Promise<void>) => {
+      const t = {};
+      return Promise.race([p, t]).then(
+        v => (v === t ? 'pending' : 'fulfilled'),
+        () => 'rejected',
+      );
+    };
+
+    // Just a short break between managing the sets of requests and active jobs
+    const createSleepPromise = () =>
+      sleep(SLEEP_TIME).then(_ => {
+        return 0;
```

Comment:
Let's use something different from 0 and 1 to be more explicit, it took me a while to understand why this `return 0`.